### DISCUSSION TITLE
Pie chart: Value mapping based on additional string field, not numerical field

### DISCUSSION THREAD

**Discussion by gnthibault** [grafana/grafana#61031] [Dashboards, Panels & Visualizations]
This is an enhancement request, as a follow up of this message:

        For the Pie Chart panel you are trying to map the label to a colour. You need to map the corresponding value of the label to the desired colour.

The visualisations represent data, so in this specific case, the colour of a slice in the PieChart should be decided by a value and not the label, as shown in my comment above, since the slice represents a value (e.g: `42.4 K` is the value represented on the Pie Chart slice and not `Meteor Mapped`, following your example above).

I will close this issue for now, since it is not a bug but more of an enhancement. If you wish, you can open a Feature Request in our Discussions section. Sadly, we will not be able to focus on this for the time being.

Thank you for opening this issue!

_Originally posted by @mdvictor in https://github.com/grafana/grafana/issues/46416#issuecomment-1099354141_

Basically, I am managing a robot driven by a statemachine, and I use the state timeline which is quite convenient, but sometimes the slices are quite short, and I built a pie chart that shows in which state did the robot spent most of the time.
This is working well, unfortunately, the vizualization does not allow to have an override that enables value matching based on regexp from the name of each pie slice (used as a legend actually).
I think this feature is very useful, especially when you put pie chart of states next to state timeline, because then you can have matching colors on both graphs.

Thank you in advance for your help

**Comment by usmangt** [comment]
Hi @gnthibault,

Thanks for opening this issue.

I will convert this issue into a Feature request as per your request (which will appear in the "Discussions" tab here on GitHub).

**Comment by gnthibault** [comment]
Ah, thank you very much @usmangt  and sorry for creating the ticket in the wrong section, I was not familiar with discussion.
I continued and tested the whole solution (which overall I am very happy with), but I still think it would be extremely useful to have a name-based "value" mapping.

To support this specific usecase, that relates to "ressource usage by state" pie-chart, I think it could be good that I share my json + a screenshot:
![dashboard_grafana_1](https://user-images.githubusercontent.com/5237397/210778193-784edce8-f3e4-4b7d-84d4-9f0a062904ac.png)
[dashboard_observatory.zip](https://github.com/grafana/grafana/files/10352061/dashboard_observatory.zip)

And in particular, the flux code (beginner version) that allowed be to build the state duration dataset:

```
import "array"

states_time_serie = from(bucket: "$bucket")
  |> range(start: v.timeRangeStart, stop:v.timeRangeStop)
  |> filter(fn: (r) => r.host == "$host")
  |> filter(fn: (r) => r._measurement == "mqtt_observatory_status")
  |> filter(fn: (r) => r._field == "state")

unique_states = states_time_serie
  |> unique()
  |> findColumn(fn: (key) => true, column: "_value")

// Define a helper function to extract a row as a record
getRow = (tables=<-, idx=0) => {
    extract = tables
        |> findRecord(fn: (key) => true, idx: idx)
    return extract
}

get_state_duration = (x) => {
  state_duration = states_time_serie
    // For each input table, elapsed() returns the same table without the first row (because there is no previous time to derive the elapsed time from) and an additional column containing the elapsed time.
    |> stateDuration(fn: (r) =>
      r._value == x,
      column: "state_duration",
      unit: 1s
      )
    // We need to sort backward in order to get positive value when doing difference on elapsed time to correspond on the current state
    //|> sort(columns: ["_time"], desc: true)
    // For each input table with n rows, difference() outputs a table with n - 1 rows.
    |> difference(columns: ["state_duration"], nonNegative: false)
    |> keep(columns: ["_time", "state_duration"])
    |> filter(fn: (r) => r["state_duration"] >= 0)
    |> sum(column: "state_duration")
    |> set(key: "state_name", value: x)
    |> getRow()
  return state_duration
}

// Computer the state duration for each state
states_durations = unique_states |> array.map(fn: get_state_duration)

// Output the array of records as a table
array.from(rows: states_durations)
```

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82816. Feel free to continue the discussion around this there. Thank you!
